### PR TITLE
Add typing for `steps/discover/fmf.py`

### PR DESCRIPTION
**Pull request by guy9050**
Blocked on StepData changes.

### CODE REVIEW

**Comment by happz** [comment]
> @happz I have one error on line 450:
> 
> ```
> tmt/steps/discover/fmf.py:450: error: On Python 3 formatting "b'abc'" with "{}" produces "b'abc'", not "abc"; use "{!r}" if this is desired behavior
> ```
> 
> Should I tell typing to ignore this line?

I can't run tests myself, but, the message basically says that there are bytes - a sequence of bytes, often mistaken for a string - used inside an f-string, and it would be rendered in an unexpected way. In this case, the return value of `re.escape(name)` is probably `bytes`, not `str`, which is probably thanks to `name` being `bytes` and not a string. Could you check what type are `name`, and `output`?

**Comment by guy9050** [comment]
> > @happz I have one error on line 450:
> > ```
> > tmt/steps/discover/fmf.py:450: error: On Python 3 formatting "b'abc'" with "{}" produces "b'abc'", not "abc"; use "{!r}" if this is desired behavior
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Should I tell typing to ignore this line?
> 
> I can't run tests myself, but, the message basically says that there are bytes - a sequence of bytes, often mistaken for a string - used inside an f-string, and it would be rendered in an unexpected way. In this case, the return value of `re.escape(name)` is probably `bytes`, not `str`, which is probably thanks to `name` being `bytes` and not a string. Could you check what type are `name`, and `output`?

So this is a really long function (declared on line 206), the function doesn't return anything. What's weird is that it's complaining about this f-string while two lines before (on line 448) there is another f-string manipulation and it doesn't complain about that.

**Comment by happz** [comment]
> > > @happz I have one error on line 450:
> > > ```
> > > tmt/steps/discover/fmf.py:450: error: On Python 3 formatting "b'abc'" with "{}" produces "b'abc'", not "abc"; use "{!r}" if this is desired behavior
> > > ```
> > >   
> > > Should I tell typing to ignore this line?
> > 
> > 
> > I can't run tests myself, but, the message basically says that there are bytes - a sequence of bytes, often mistaken for a string - used inside an f-string, and it would be rendered in an unexpected way. In this case, the return value of `re.escape(name)` is probably `bytes`, not `str`, which is probably thanks to `name` being `bytes` and not a string. Could you check what type are `name`, and `output`?
> 
> So this is a really long function (declared on line 206), the function doesn't return anything. What's weird is that it's complaining about this f-string while two lines before (on line 448) there is another f-string manipulation and it doesn't complain about that.

The problem is not the f-string per se, that is completely fine. What the message is trying to tell you is that one of the variables used in the f-string is of `bytes` type, which may have unexpected side-effects. For example:

```python
>>> foo = 'bar'
>>> f'{foo}'
'bar'
>>> foo = b'bar'
>>> f'{foo}'
"b'bar'"
>>> 
```

See the little `b` sneaking into the formatted string - this is certainly not the outcome I was hoping to get. So, the message is pointing at one of the inputs of the f-string, not the f-string itself.

But, there seem to be no `bytes` around:

```python
            modified = set(
                f"^/{re.escape(name)}"
                for name in map(os.path.dirname, output.split('\n')) if name)
```

`output` is a string, therefore `.split()` will return `List[str]`, and `os.path.dirname`, `map` and `re.escape` calls should preserve `str` input by returning `str` output. `os.path.dirname` and `re.escape` annotate their input parameters with `AnyStr` type, which si a helper type that boils down to `Union[str, bytes]` in this case. The information that input is a string, and therefore output will also `str` and not `bytes` is "lost in translation", somewhere between `map()` and the loop. So, to re-introduce the correct type, you can do this:

```python
                f"^/{re.escape(cast(str, name))}"
```

This will tell mypy that `re.escape` input is a string, not `bytes`, and mypy will therefore infer the output of `re.escape` will be `str` too, removing the possibility of `bytes` out of the picture.

**Comment by guy9050** [comment]
@happz @thrix please review

File: tmt/steps/discover/fmf.py
Comment (lukaszachy) [inline tmt/steps/discover/fmf.py:200]: Why is this 'Any' if it has to return bool? Because no all parent classes are typed yet?

File: tmt/steps/discover/fmf.py
Comment (lukaszachy) [inline tmt/steps/discover/fmf.py:229]: It has to return Str

File: tmt/steps/discover/fmf.py
Comment (janhavlin) [inline tmt/steps/discover/fmf.py:206]: Seems like this method doesn't return anything, so `def go(self) -> None:`?

File: tmt/steps/discover/fmf.py
Comment (janhavlin) [inline tmt/steps/discover/fmf.py:123]: Should return `List[ClickOptionDecoratorType]` which is implemented in #1304.

File: tmt/steps/discover/fmf.py
Comment (guy9050) [reply tmt/steps/discover/fmf.py:229]: how do I overcome tmt/steps/discover/fmf.py:230: error: Returning Any from function declared to return "str" ?

File: tmt/steps/discover/fmf.py
Comment (guy9050) [reply tmt/steps/discover/fmf.py:200]: mypy says it returns `Any` but I agree it should be `bool`, maybe some explicit declaration should be added 
`tmt/steps/discover/fmf.py:204: error: Returning Any from function declared to return "bool"`

**Comment by guy9050** [comment]
blocked on https://github.com/teemtee/tmt/pull/1304

File: tmt/steps/discover/fmf.py
Comment (guy9050) [reply tmt/steps/discover/fmf.py]: thanks! wasn't aware of that object

File: tmt/steps/discover/fmf.py
Comment (guy9050) [reply tmt/steps/discover/fmf.py:201]: fixed, thx

Comment (guy9050) [reply tmt/steps/discover/fmf.py]: fixed, please approve

File: tmt/steps/discover/fmf.py
Comment (janhavlin) [inline tmt/steps/discover/fmf.py:18]: Also blocked on https://github.com/teemtee/tmt/pull/1333 due to this `# type: ignore`, right? It is also ready for review so we should focus on that PR first.

File: tmt/steps/discover/fmf.py
Comment (janhavlin) [inline tmt/steps/discover/fmf.py:184]: This `cast()` should be unnecessary once #1333 is merged.

File: tmt/steps/discover/fmf.py
Comment (guy9050) [inline tmt/steps/discover/fmf.py:18]:
```suggestion
# TODO: get rid of `ignore` once superclass is no longer `Any`
class DiscoverFmf(tmt.steps.discover.DiscoverPlugin):  # type: ignore[misc]
```

File: tmt/steps/discover/fmf.py
Comment (thrix) [inline tmt/steps/discover/fmf.py:452]: @guy9050 could we move the cast rather to output assignement?

**Review by thrix — CHANGES_REQUESTED** [review]
@guy9050 found one more thing

File: tmt/steps/discover/fmf.py
Comment (guy9050) [reply tmt/steps/discover/fmf.py:19]: That's not exactly blocking it, it can still be merged and there are a lot of places with such things that are already merged, i.e https://github.com/teemtee/tmt/blob/main/tmt/steps/provision/artemis.py#L234 .
I added a comment to remove it once its superclass is typed

File: tmt/steps/discover/fmf.py
Comment (guy9050) [reply tmt/steps/discover/fmf.py:185]: Yes. But it's not merged yet.

File: tmt/steps/discover/fmf.py
Comment (thrix) [reply tmt/steps/discover/fmf.py:451]: Ack, sorry missed it, we will anyway revisit all casts.

**Review by thrix — APPROVED** [review]
LGTM

File: tmt/steps/discover/fmf.py
Comment (thrix) [reply tmt/steps/discover/fmf.py:184]: Actually we have a workaround for this now

**Comment by thrix** [comment]
@guy9050 I squashed the commits here and applied the correct solution we agreed on today. It fails because it needs changes introduced in #1333. Once that is merged, it will start passing.

**Comment by guy9050** [comment]
> @guy9050 I squashed the commits here and applied the correct solution we agreed on today. It fails because it needs changes introduced in #1333. Once that is merged, it will start passing.

@thrix ack. thanks.

**Comment by thrix** [comment]
@guy9050 please rebase, the blocking PR of this PR is now merged \o/

**Comment by thrix** [comment]
Moved to 1.17, this needs StepData changes

**Comment by psss** [comment]
Rebased on the latest `main`, added a few minor changes in dcf51ad7. @thrix, @happz, please, have a look.

**Review by psss — APPROVED** [review]
Looks good, added just a few minor changes.

File: tmt/steps/discover/fmf.py
Comment (happz) [inline tmt/steps/discover/fmf.py:245]: This `cast()` is suspicious. `self.run()` returns `tmt.utils.CommandOutput`, and its `.stdout` attribute is `Optional[str]`. So `str` is already there, but it may be `Optional`, i.e. `stdout is None` would be true. `cast()` is effectively masking this fact, pretending `stdout` is *always* not `None`. `output = self.run(...)` followed by `assert out.stdout is not None` should be enough to "prove" we can safely return `output.stdout`, with no `cast()` involved. Types match, it's the optionality that's preventing simple `return self.run(...).stdout`.

**Review by happz — APPROVED** [review]

**Pull request merged**
